File: tests/game_objects/systems/test_inventory.py
"""Tests all classes and functions in game_objects/systems/inventory.py."""
from __future__ import annotations

# Pip
import pytest

# Custom
from hades.game_objects.components import Inventory
from hades.game_objects.registry import Registry, RegistryError
from hades.game_objects.systems.inventory import InventorySpaceError, InventorySystem

__all__ = ()


@pytest.fixture()
def registry() -> Registry:
    """Create a registry for use in testing.

    Returns:
        The registry for use in testing.
    """
    return Registry()


@pytest.fixture()
def inventory_system(registry: Registry) -> InventorySystem:
    """Create an inventory system for use in testing.

    Args:
        registry: The registry for use in testing.

    Returns:
        The inventory system for use in testing.
    """
    registry.create_game_object(Inventory(3, 6))
    inventory_system = InventorySystem(registry)
    registry.add_system(inventory_system)
    return inventory_system


def test_raise_inventory_space_error_full() -> None:
    """Test that InventorySpaceError is raised correctly when full."""
    with pytest.raises(
        expected_exception=InventorySpaceError,
        match="The inventory is full.",
    ):
        raise InventorySpaceError(full=True)


def test_raise_inventory_space_error_empty() -> None:
    """Test that InventorySpaceError is raised correctly when empty."""
    with pytest.raises(
        expected_exception=InventorySpaceError,
        match="The inventory is empty.",
    ):
        raise InventorySpaceError(full=False)


def test_inventory_system_init(inventory_system: InventorySystem) -> None:
    """Test that the inventory system is initialised correctly.

    Args:
        inventory_system: The inventory system for use in testing.
    """
    assert (
        repr(inventory_system)
        == "<InventorySystem (Description=`Provides facilities to manipulate inventory"
        " components.`)>"
    )


def test_inventory_system_add_item_to_inventory_valid(
    inventory_system: InventorySystem,
) -> None:
    """Test that a valid item is added to the inventory correctly.

    Args:
        inventory_system: The inventory system for use in testing.
    """
    inventory_system.add_item_to_inventory(0, 50)
    assert inventory_system.registry.get_component_for_game_object(
        0,
        Inventory,
    ).inventory == [50]


def test_inventory_system_add_item_to_inventory_zero_size(
    inventory_system: InventorySystem,
) -> None:
    """Test that a valid item is not added to a zero size inventory.

    Args:
        inventory_system: The inventory system for use in testing.
    """
    inventory_system.registry.get_component_for_game_object(0, Inventory).width = 0
    with pytest.raises(expected_exception=InventorySpaceError):
        inventory_system.add_item_to_inventory(0, 50)


def test_inventory_system_add_item_to_inventory_invalid_game_object_id(
    inventory_system: InventorySystem,
) -> None:
    """Test that an exception is raised if an invalid game object ID is provided.

    Args:
        inventory_system: The inventory system for use in testing.
    """
    with pytest.raises(
        expected_exception=RegistryError,
        match="The game object ID `-1` is not registered with the registry.",
    ):
        inventory_system.add_item_to_inventory(-1, 50)


def test_inventory_system_remove_item_from_inventory_valid(
    inventory_system: InventorySystem,
) -> None:
    """Test that a valid item is removed from the inventory correctly.

    Args:
        inventory_system: The inventory system for use in testing.
    """
    inventory_system.add_item_to_inventory(0, 1)
    inventory_system.add_item_to_inventory(0, 7)
    inventory_system.add_item_to_inventory(0, 4)
    assert inventory_system.remove_item_from_inventory(0, 1) == 7
    assert inventory_system.registry.get_component_for_game_object(
        0,
        Inventory,
    ).inventory == [1, 4]


def test_inventory_system_remove_item_from_inventory_large_index(
    inventory_system: InventorySystem,
) -> None:
    """Test that an exception is raised if a larger index is provided.

    Args:
        inventory_system: The inventory system for use in testing.
    """
    inventory_system.add_item_to_inventory(0, 5)
    inventory_system.add_item_to_inventory(0, 10)
    inventory_system.add_item_to_inventory(0, 50)
    with pytest.raises(expected_exception=InventorySpaceError):
        inventory_system.remove_item_from_inventory(0, 10)


def test_inventory_system_remove_item_from_inventory_invalid_game_object_id(
    inventory_system: InventorySystem,
) -> None:
    """Test that an exception is raised if an invalid game object ID is provided.

    Args:
        inventory_system: The inventory system for use in testing.
    """
    with pytest.raises(
        expected_exception=RegistryError,
        match="The game object ID `-1` is not registered with the registry.",
    ):
        inventory_system.remove_item_from_inventory(-1, 0)

File: hades_extensions/test/game_objects/systems/test_inventory.cpp
// External includes
#include "gtest/gtest.h"

// Custom includes
#include "macros.hpp"
#include "game_objects/systems/inventory.hpp"

// ----- FIXTURES ------------------------------
/// Implements the Inventory fixture for the game_objects/systems/inventory.hpp tests.
class InventoryFixture : public testing::Test {
 protected:
  /// The registry that manages the game objects, components, and systems.
  Registry registry{};

  /// Set up the fixture for the tests.
  void SetUp() override {
    std::vector<std::unique_ptr<ComponentBase>> components;
    components.push_back(std::make_unique<Inventory>(3, 6));
    registry.create_game_object(false, std::move(components));
    registry.add_system(std::make_shared<InventorySystem>(registry));
  }

  /// Get the inventory system from the registry.
  ///
  /// @return The inventory system.
  std::shared_ptr<InventorySystem> get_inventory_system() {
    return registry.find_system<InventorySystem>();
  }
};

// ----- TESTS ----------------------------------
/// Test that InventorySpaceError is raised correctly when full.
TEST(Tests, TestThrowInventorySpaceErrorFull) {
  ASSERT_THROW_MESSAGE(throw InventorySpaceException(true), InventorySpaceException, "The inventory is full.")
}

/// Test that InventorySpaceError is raised correctly when empty.
TEST(Tests, TestThrowInventorySpaceErrorEmpty) {
  ASSERT_THROW_MESSAGE(throw InventorySpaceException(false), InventorySpaceException, "The inventory is empty.")
}

/// Test that a valid item is added to the inventory correctly.
TEST_F(InventoryFixture, TestInventorySystemAddItemToInventoryValid) {
  get_inventory_system()->add_item_to_inventory(0, 50);
  ASSERT_EQ(registry.get_component<Inventory>(0)->items, std::vector<int>{50});
}

/// Test that a valid item is not added to a zero size inventory.
TEST_F(InventoryFixture, TestInventorySystemAddItemToInventoryZeroSize) {
  registry.get_component<Inventory>(0)->width = 0;
  ASSERT_THROW_MESSAGE(get_inventory_system()->add_item_to_inventory(0, 50),
                       InventorySpaceException,
                       "The inventory is full.")
}

/// Test that an exception is raised if an invalid game object ID is provided.
TEST_F(InventoryFixture, TestInventorySystemAddItemToInventoryInvalidGameObjectID) {
  ASSERT_THROW_MESSAGE(get_inventory_system()->add_item_to_inventory(-1, 50),
                       RegistryException,
                       "The game object `-1` is not registered with the registry.")
}

/// Test that a valid item is removed from the inventory correctly.
TEST_F(InventoryFixture, TestInventorySystemRemoveItemFromInventoryValid) {
  std::vector<int> result{1, 4};
  get_inventory_system()->add_item_to_inventory(0, 1);
  get_inventory_system()->add_item_to_inventory(0, 7);
  get_inventory_system()->add_item_to_inventory(0, 4);
  ASSERT_EQ(get_inventory_system()->remove_item_from_inventory(0, 1), 7);
  ASSERT_EQ(registry.get_component<Inventory>(0)->items, result);
}

/// Test that an exception is raised if a larger index is provided.
TEST_F(InventoryFixture, TestInventorySystemRemoveItemFromInventoryLargeIndex) {
  get_inventory_system()->add_item_to_inventory(0, 5);
  get_inventory_system()->add_item_to_inventory(0, 10);
  get_inventory_system()->add_item_to_inventory(0, 50);
  ASSERT_THROW_MESSAGE(get_inventory_system()->remove_item_from_inventory(0, 10),
                       InventorySpaceException,
                       "The index is out of range.")
}

/// Test that an exception is raised if an invalid game object ID is provided.
TEST_F(InventoryFixture, TestInventorySystemRemoveItemFromInventoryInvalidGameObjectID) {
  ASSERT_THROW_MESSAGE(get_inventory_system()->remove_item_from_inventory(-1, 0),
                       RegistryException,
                       "The game object `-1` is not registered with the registry.")
}
